### PR TITLE
Template dir option

### LINKED ISSUE #714: Would you be open to a PR that adds in an option to use alternative templates?
Hello! 

We ([REVSYS](https://www.revsys.com)) have a client, a fellow Open Source project, that would like to use gcovr for their coverage reports.  However, they would like to heavily adjust the styling to better suit their preferences and that of their project's website.  

Would you all be open to a PR, if I did the work, to allow gcovr to support an `--html` output option but also honor a `--template-dir` flag to point at a different set of jinja2 templates?

### PR DESCRIPTION
**Pull request by frankwiles**
This PR adds a `--template-dir` option to the CLI which allows the use of a directory of alternate Jinja2 templates for users who want to heavily customize their HTML output but don't want to fork gcovr to do so.  

I believe I've covered all the bases here, README, docs, etc but let me know if I've missed anything you would like to see. 

Closes #714

### CODE REVIEW

**Comment by vinniefalco** [comment]
This is quite frankly the most beautiful pull request I have ever seen in my life

File: gcovr/configuration.py
Comment (Spacetown) [inline gcovr/configuration.py:920]: Use a prefix since this is hor HTML only.
```suggestion
        "html_template_dir",
        ["--html_template-dir"],
```

File: gcovr/configuration.py
Comment (Spacetown) [inline gcovr/configuration.py:924]: Relative? I think it can also be absolute.

File: gcovr/writer/html/__init__.py
Comment (Spacetown) [inline gcovr/writer/html/__init__.py:80]: Why creating a loader if it's immediately overwritten?
```suggestion
    if options.template_dir is None:
        loader = PackageLoader("gcovr.writer.html")
    else:
        loader = FileSystemLoader(options.template_dir)
```

File: gcovr/writer/html/__init__.py
Comment (Spacetown) [inline gcovr/writer/html/__init__.py:83]: Is there also a list of loaders possible? Then the user can override a specific template but reusing the others.

**Review by Spacetown — CHANGES_REQUESTED** [review]
I'm restructuring the writers in a separate PR which will be merged today and you need to rebase.
Please add yourself to the cCONTRIBUTORS.rst and the issue to the CHANGELOG.rst.

File: gcovr/configuration.py
Comment (frankwiles) [reply gcovr/configuration.py:920]: @Spacetown did you mean `--html-template-dir` here ? Underscore seems weird.

File: gcovr/writer/html/__init__.py
Comment (frankwiles) [reply gcovr/writer/html/__init__.py:90]: Yeah that's a good idea, Jinja2 does have a `ChoiceLoader` which I just added into the mix. This way a user only has to override the templates they care to change and don't have to copy them all into a directory.

**Comment by frankwiles** [comment]
I'm confused on this, I've rebased from upstream master and my local branch is clean of these conflicts. But the PR is still showing them.

File: gcovr/configuration.py
Comment (Spacetown) [reply gcovr/configuration.py]: The option always uses `-` where the key uses `_`.

File: gcovr/writer/html/__init__.py
Comment (Spacetown) [reply gcovr/writer/html/__init__.py:90]: Can you also update the docs here?

**Comment by Spacetown** [comment]
> I'm confused on this, I've rebased from upstream master and my local branch is clean of these conflicts. But the PR is still showing them.
> 
> I did end up using VS Code merge tool (for the first time) so I suppose it's possible I messed something up but I have no conflict markers in these files.

There was no merge from the master branch to yours. After merging there where two conflicts I need to resolve.

**Comment by Spacetown** [comment]
@frankwiles Please can you allow maintainers to push on your branch?

**Comment by frankwiles** [comment]
@Spacetown Sure, but I'm not finding where one does that in the UI.  All I found was this page which says to do it on the PR list but I'm not seeing the option anywhere: 

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Honestly didn't know Github had that feature which is pretty neat.

**Comment by Spacetown** [comment]
> @Spacetown Sure, but I'm not finding where one does that in the UI. All I found was this page which says to do it on the PR list but I'm not seeing the option anywhere:
> 
> https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork
> 
> Honestly didn't know Github had that feature which is pretty neat.
Bildschirm­foto 2023-04-07 um 20 47 28

This is a checkbox when opening a PR. For already opened PRs it's on the right list at the end:
Bildschirm­foto 2023-04-07 um 20 47 28

**Comment by frankwiles** [comment]
CleanShot 2023-04-07 at 14 04 08

Hmmm I'm not seeing an option there for that. Is this maybe because I forked this to an org rather than an individual?

**Comment by Spacetown** [comment]
That's possible. Please merge the main branch to your branch and solve the conflicts. There won't be a changed file but the merge information needs to be present.

File: README.rst
Comment (Spacetown) [inline README.rst:38]: The `html-` is missing here. I'm not sure if this option shall be mentioned in the overview since it's only a customization for the `--html` or `--html-details`.
```suggestion
   * - |abstract-option-html-template-dir|
```

File: doc/source/output/html.rst
Comment (Spacetown) [inline doc/source/output/html.rst:71]: I would rewrite this part a little bit:
```suggestion
The :option:`--html-template-dir<gcovr --html-template-dir>` option allows you to use an
alternate directory with Jinja2 templates to fully customize
the HTML output in detail.  It is advised that you start by copying the part you want to modify from
`the existing templates <https://github.com/gcovr/gcovr/tree/master/gcovr/writer/html/templates>`_
. Keep in mind the template context data that is passed to these templates *MAY* change
between release versions.

Note that you do not have to copy every single template and can copy and edit only the
templates you wish to customize.

```

Can you also change line 46 to the following content? I saw that `--html-details` is used twice in the sentence.
```
Note that the :option:`--html<gcovr --html>` and
```

**Comment by frankwiles** [comment]
This is really weird, if I try to merge it I get THE EXACT same merge conflicts I got (and have resolved a couple of times now) rebasing.  Which are the two files it says it has conflicts on.  

Any objections to me re-forking on my personal account, dropping in the changes and creating a new PR? I think this one is somehow cursed! 😄

**Review by Spacetown — COMMENTED** [review]
Some suggestions. After the merge from master to your branch there should be less changes to review.

**Comment by Spacetown** [comment]
We'll get it working here. Can you clone from my repo (https://github.com/Spacetown/gcovr/tree/template-dir-option), rest your branch to this one and push from your side?

**Comment by frankwiles** [comment]
> This is really weird, if I try to merge it I get THE EXACT same merge conflicts I got (and have resolved a couple of times now) rebasing. Which are the two files it says it has conflicts on.
> 
> Any objections to me re-forking on my personal account, dropping in the changes and creating a new PR? I think this one is somehow cursed! 😄

Nevermind! I managed to realize what you meant on the merge and get it sorted.

File: gcovr/formats/html/__init__.py
Comment (Spacetown) [inline gcovr/formats/html/__init__.py:36]: In this file the manual merge is wrong, we've now the copyright notice and the class twice.

Comment (frankwiles) [reply gcovr/formats/html/__init__.py:36]: Bah, I really shouldn't have used VS Code's merge tool without understanding it better it seems! Fixing.

File: gcovr/formats/html/write.py
Comment (Spacetown) [inline gcovr/formats/html/write.py:138]: I think with this we can remove the option `--html-css` because the user can provide one template directory where a `style.css` is present two have the same functionality.

Comment (frankwiles) [reply gcovr/formats/html/write.py:138]: I definitely see your point, but wouldn't that confuse existing users who might only be doing CSS adjustments? 

To be clear, I'm happy to do the removal, just with it being such a small footprint of a feature I don't think the overlap of functionality is that bad.

Comment (Spacetown) [reply gcovr/formats/html/write.py:138]: Since the new option will imply a fixed name I think it's better to keep the current behavior.

File: gcovr/tests/test_args.py
Comment (Spacetown) [inline gcovr/tests/test_args.py:164]: The naming doesn't fit here since this isn't an output option.

Comment (frankwiles) [reply gcovr/tests/test_args.py]: Ah yeah and it wasn't even really testing something useful.  I've updated it.

File: doc/source/output/html.rst
Comment (Spacetown) [reply doc/source/output/html.rst:71]: This isn't resolved yet. The first sentence doesn't make sense and the recommendation should be to only overwrite what's needed and not all templates.

**Comment by Spacetown** [comment]
@frankwiles You also need to merge the master into your branch to get the CI running.

**Comment by vinniefalco** [comment]
image

**Comment by vinniefalco** [comment]
image

**Review by Spacetown — COMMENTED** [review]
Documentation still neeeds to be updated.

**Comment by frankwiles** [comment]
Apologies on the delay on this @Spacetown I think this accomplishes the change, but let me know if you want it worded differently.

**Comment by Spacetown** [comment]
I'm missing a sentence which highlights the inheritance from the base directory. Something like `If a template is missing in this directory the original one is used.`.
Also a test is missing where e.g. `base.html` and `directory_page.summary.html` is added to the given template directory.

**Comment by frankwiles** [comment]
Ok I added a test that ensures the internal _mechanics_ of template overriding behaves properly, without resorting to testing various HTML differences between the two which I hope is what you are after.

File: gcovr/formats/html/write.py
Comment (Spacetown) [inline gcovr/formats/html/write.py:87]: This print must be removed.

**Review by Spacetown — COMMENTED** [review]
Documentation is now OK.
For the test I suggest to copy the test `gcovr/tests/html-css` to `gcovr/tests/html-template-dir`, copy some default templates to the config directory and use this directory as template directory.

File: gcovr/formats/html/write.py
Comment (frankwiles) [reply gcovr/formats/html/write.py]: Oh jeez, that's embarrassing! Was rushing to get it in before I had to catch a flight.

Comment (Spacetown) [reply gcovr/formats/html/write.py]: No problem.

**Comment by frankwiles** [comment]
Ok I did the base template and directory summary and just put in sentinel values we can easily watch for.

**Comment by Spacetown** [comment]
Can you fix the failed test?

**Comment by frankwiles** [comment]
@Spacetown sorry I got distracted by other things.  I ran this locally and there are still some small differences in the HTML output, but the default templates haven't changed.  I essentially copied the `html-css` test and it's styles, etc. just overriding the contents of a `<th>` and the page title to show the template fall through is working.  

Hoping seeing the output in CI leads me to what else is going wrong as I can't run these tests locally (at least not meaningfully)

**Comment by Spacetown** [comment]
The different files will be attached as Artifacts.

**Comment by frankwiles** [comment]
Ok, I'm starting to run out of ideas as to why this is happening.  If I run `pytest gcovr/tests/test_html_generator.py` locally (removing the timeout option in setup.cfg as it seems to be missing in this version of pytest) the test passes just fine. 

If I purposefully insert an `assert False` at the end you can see with these debugging prints it does as expected.  Using the option works as expected when running `gcovr` itself, but when running `nox` it fails.  

Is it perhaps something related to the nox environment I'm not thinking of? Something that would perhaps cache or reuse something in a way I'm not realizing? 

CleanShot 2023-06-28 at 07 10 14@2x

**Comment by Spacetown** [comment]
Sorry, the next few weeks I can't check this. I'll investigate after my trip.

**Comment by frankwiles** [comment]
@Spacetown no worries, thanks! Enjoy your trip!

**Comment by frankwiles** [comment]
@Spacetown hi, just wanted to check in and see if you had a chance to look at this at all.  So confused as to why it passes locally and doesn't in CI.

**Comment by Spacetown** [comment]
I was 5 weeks on vacation and now I first want to fix the test on the Windows runners.

**Comment by frankwiles** [comment]
That's great @Spacetown ! I'm jealous of so much vacation time! 😄

File: gcovr/tests/test_html_generator.py
Comment (Spacetown) [inline gcovr/tests/test_html_generator.py:93]:
```suggestion
    template_dir: pathlib.Path = tmp_path_factory.mktemp("alt_templates", numbered=True)
    base_template = template_dir / "base.html"
    directory_template = template_dir / "directory_page.summary.html"

    # Write some content we can spot in the templates
    base_template.write_text("NEW_BASE_TEMPLATE")
    directory_template.write_text("NEW_DIRECTORY_TEMPLATE")

    return template_dir
```

File: gcovr/formats/html/write.py
Comment (Spacetown) [inline gcovr/formats/html/write.py:81]: I don't understand the test failure in the CI. Even if I put a exception here it isn't raised with the injection class and I don't know why.

I suggest to remove this test since you've also added a makefile test.

**Review by Spacetown — CHANGES_REQUESTED** [review]

**Comment by frankwiles** [comment]
Glad it wasn't just me who was baffled on that one! Ok, test removed and rebased with master.

**Comment by Spacetown** [comment]
@frankwiles I want to finish this but without beeing able to push to the branch it's difficult. Can you please allow my account to push in general to your fork?

**Comment by frankwiles** [comment]
@Spacetown sure thing, just sent you an invite to collaborate on it

**Comment by Spacetown** [comment]
Pushing from my local machine still doesn't work.
In Codespaces or the conflict solver I can commit now.

**Comment by Spacetown** [comment]
@frankwiles Anything to add or check from your side?

**Comment by frankwiles** [comment]
No, everything looks good to me.  Thanks for all your help on this @Spacetown !

**Review by Spacetown — APPROVED** [review]
LGFM

**Pull request merged**
